File: 0155-min-stack/0155-min-stack.rb
class Node
  attr_accessor :val, :min
  
=begin
  :type val: Integer
  :type min: Integer
=end
  def initialize(val)
    @val = val
    @min = nil
  end
  
end

class MinStack
  def initialize()
    @stack = []
  end


=begin
  :type val: Integer
  :rtype: Void
=end
  def push(val)
    new_node = Node.new(val)
    if @stack.empty? || get_min() > val
      new_node.min = val
    else
      new_node.min = get_min()
    end
    @stack << new_node
  end


=begin
  :rtype: Void
=end
  def pop()
    @stack.pop
    nil
  end


=begin
  :rtype: Integer
=end
  def top()
    @stack[-1].val
  end


=begin
  :rtype: Integer
=end
  def get_min()
    @stack[-1].min
  end


end

# Your MinStack object will be instantiated and called as such:
# obj = MinStack.new()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.get_min()

File: 0155-min-stack/0155-min-stack.java
public class Node {
  protected int val;
  protected int min;
  
  public Node(int val) {
    this.val = val;
  }
}

class MinStack {
  protected List<Node> stack;

  public MinStack() {
    this.stack = new ArrayList<Node>();
  }

  public void push(int val) {
    Node newNode = new Node(val);
    if (this.stack.isEmpty() || getMin() > val) {
      newNode.min = val;
    } else {
      newNode.min = getMin();
    }
    this.stack.add(newNode);
  }

  public void pop() {
    this.stack.remove(this.stack.size() - 1);
  }

  public int top() {
    return this.stack.get(this.stack.size() - 1).val;
  }

  public int getMin() {
    return this.stack.get(this.stack.size() - 1).min;
  }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */